File: LeetCode/46.permutations.cpp
/*
 * @lc app=leetcode id=46 lang=cpp
 *
 * [46] Permutations
 */

// @lc code=start
class Solution {
public:
   vector<vector<int>>ans;
    

    vector<vector<int>> permute(vector<int>& nums)
    {
        permutation(nums,0, nums.size()-1);

        return ans;
    }

    void permutation(vector<int>& nums, int l, int r)
    {
        if(l==r)
        {
            ans.push_back(nums);
            return;
        }

        for(int i=l;i<=r;i++)
        {
            swap(nums[l], nums[i]);
            permutation(nums, l+1, r);
            swap(nums[l], nums[i]);
        }
    }




    
};
// @lc code=end

File: LeetCode/46.permutations.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    List<List<Integer>> ans;

    void swap(int i, int j, int[] nums) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    void rec(int[] nums, int ind) {
        if (ind == nums.length - 1) {
            List<Integer> ls = new ArrayList<>();
            for (int it : nums) {
                ls.add(it);
            }
            ans.add(ls);
            return;
        }

        for (int i = ind; i < nums.length; i++) {
            swap(i, ind, nums);
            rec(nums, ind + 1);  // Fix: Pass 'ind + 1' instead of 'i + 1'
            swap(i, ind, nums);
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        ans = new ArrayList<>();
        rec(nums, 0);
        return ans;
    }
}
